patroni/postgresql/callback_executor.py

```diff
@@ -1,40 +1,44 @@
-import logging
-import subprocess
 from threading import Event, Lock, Thread
+import subprocess
+import logging
+import time
+import queue
 
 logger = logging.getLogger(__name__)
 
-
 class CallbackExecutor(Thread):
+    '''
+    Execution thread that queues commands and executes them in order.
+    Guarantees that commands are executed in the order they are received via
+    FIFO queue.
+    '''
 
-    def __init__(self):
+    def __init__(self, executor=subprocess.Popen):
         super(CallbackExecutor, self).__init__()
         self.daemon = True
-        self._lock = Lock()
-        self._cmd = None
-        self._process = None
-        self._callback_event = Event()
+        self._cmds_queue = queue.Queue()
+        self._executor = executor
         self.start()
 
     def call(self, cmd):
-        with self._lock:
-            if self._process and self._process.poll() is None:
-                try:
-                    self._process.kill()
-                    logger.warning('Killed the old callback process because it was still running: %s', self._cmd)
-                except OSError:
-                    logger.exception('Failed to kill the old callback')
-        self._cmd = cmd
-        self._callback_event.set()
+      # Queue up the command
+      self._cmds_queue.put(cmd)
 
     def run(self):
         while True:
-            self._callback_event.wait()
-            self._callback_event.clear()
-            with self._lock:
-                try:
-                    self._process = subprocess.Popen(self._cmd, close_fds=True)
-                except Exception:
-                    logger.exception('Failed to execute %s',  self._cmd)
-                    continue
-            self._process.wait()
+            logger.info("Callback sleeping until new command is queued")
+            cmd = self._cmds_queue.get()
+            cmd_str = ' '.join(cmd)
+            logger.info("Callback executor waking to call {}".format(cmd_str))
```

Comment:
The logging module already supports [string formatting](https://docs.python.org/2/library/stdtypes.html#string-formatting) and it is preferable over the `str.format`.